scheduler/scheduler.go

```diff
@@ -132,70 +133,70 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
 	// loop over the distros, spawning up the appropriate number of hosts
 	// for each distro
 	hostsSpawnedPerDistro := make(map[string][]host.Host)
+
 	for distroId, numHostsToSpawn := range newHostsNeeded {
 		distroStartTime := time.Now()
 
 		if numHostsToSpawn == 0 {
 			continue
 		}
 
+		if ctx.Err() != nil {
+			return nil, errors.New("scheduling run canceled.")
+		}
+
+		d, err := distro.FindOne(distro.ById(distroId))
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to find distro %s", distroId)
+		}
+
 		hostsSpawnedPerDistro[distroId] = make([]host.Host, 0, numHostsToSpawn)
-		for i := 0; i < numHostsToSpawn; i++ {
-			if ctx.Err() != nil {
-				return nil, errors.New("scheduling run canceled.")
+		// if distro can have containers, check if there are enough parents to hold
+		// new containers
+		if d.MaxContainers > 0 {
+			currentParents, err := host.FindAllRunningParents()
+			if err != nil {
+				return nil, errors.Wrap(err, "could not find running parents")
 			}
-
-			d, err := distro.FindOne(distro.ById(distroId))
+			existingContainers, err := host.FindAllRunningContainers()
 			if err != nil {
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":  distroId,
-					"runner":  RunnerName,
-					"message": "failed to find distro",
-				}))
-				continue
+				return nil, errors.Wrap(err, "could not find running containers")
 			}
 
-			allDistroHosts, err := host.Find(host.ByDistroId(distroId))
-			if err != nil {
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":  distroId,
-					"runner":  RunnerName,
-					"message": "problem getting hosts for distro",
-				}))
+			numNewParents := numNewParentsNeeded(len(currentParents), numHostsToSpawn,
+				len(existingContainers), d)
 
-				continue
-			}
+			// only want to spawn amount of parents allowed based on pool size
+			numNewParentsToSpawn := parentCapacity(d, numNewParents, len(currentParents), len(existingContainers), numHostsToSpawn)
+			// if there are parents to spawn, do not spawn containers in the same run
+			if numNewParentsToSpawn > 0 {
```

Comment:
I think this will be too inefficient. Couldn't we spawn containers on parents that are already up, and also spawn more parents if needed?